### PR TITLE
Added code to differenciate the Krusha when doing multi-Krusha runs

### PR DESCRIPTION
**Pull request by Sir-Chili**
This code was made after trying to do a rando with both 5 Krushas and Switchsanity, and struggling as I had to manually check the randomized switches to see what they were.

Had to move a lot of reserved code sections to fit the pictures so I could use the same strategy used for the first one, but made sure the references pointed to the new location.

This was tested in a completed Rando to make sure everything works fine.

Please let me know if there is anything!

### CODE REVIEW

File: randomizer/Patching/CosmeticColors.py
Comment (theballaam96) [inline randomizer/Patching/CosmeticColors.py:2560]: Replace this with a `getImageFile()` call. Use `getBonusSkinOffset()` to get the file index that's needing to be used for the file index argument. You'll then want to change the BinaryIO reads to reading from a bytearray

File: randomizer/Patching/CosmeticColors.py
Comment (theballaam96) [inline randomizer/Patching/CosmeticColors.py:2557]: Since the krusha heads now are being based on color, I'd add a check here for colorblind mode being enabled. If it is, then return out of the function early to prevent the heads being overwritten.

File: randomizer/Patching/CosmeticColors.py
Comment (theballaam96) [inline randomizer/Patching/CosmeticColors.py:3587]: Revert this

File: randomizer/Patching/CosmeticColors.py
Comment (theballaam96) [inline randomizer/Patching/CosmeticColors.py:4210]: Revert this

File: randomizer/Patching/FairyPlacer.py
Comment (theballaam96) [inline randomizer/Patching/FairyPlacer.py:1]: Revert

File: randomizer/Patching/ItemRando.py
Comment (theballaam96) [inline randomizer/Patching/ItemRando.py:1]: Revert

File: randomizer/Patching/UpdateHints.py
Comment (theballaam96) [inline randomizer/Patching/UpdateHints.py:1]: Revert the changes here

**Review by theballaam96 — CHANGES_REQUESTED** [review]
Overall, I do like the idea behind this change and think it's worth pursuing. The shifting of static locations does make me nervous, so I've suggested a method of doing this which doesn't revolve around shifting static attributes

File: randomizer/Patching/CosmeticColors.py
Comment (theballaam96) [reply randomizer/Patching/CosmeticColors.py:2560]: Update, had a little more of a think on this. getImageFile returns a PIL Image rather than a bytes object. So ignore the BinaryIO segment that. use `writeColorImageToROM()` instead

File: base-hack/Build/build.py
Comment (Sir-Chili) [reply base-hack/Build/build.py:1]: Completely fair. I'll revert the changes, then change the code replacing the Kongs' heads with the new system. Reverting changes done in this file.

File: base-hack/Build/recompute_overlays.py
Comment (Sir-Chili) [reply base-hack/Build/recompute_overlays.py:1]: Reverted

File: base-hack/include/ultra64/os.h
Comment (Sir-Chili) [reply base-hack/include/ultra64/os.h:1]: Reverted

File: base-hack/include2/PR/os_tlb.h
Comment (Sir-Chili) [reply base-hack/include2/PR/os_tlb.h:1]: Reverted

File: base-hack/src/initialization/init.c
Comment (Sir-Chili) [reply base-hack/src/initialization/init.c:1]: Reverted

File: base-hack/src/item rando/enemy_items.c
Comment (Sir-Chili) [reply base-hack/src/item rando/enemy_items.c:1]: Reverted

File: base-hack/src/main.c
Comment (Sir-Chili) [reply base-hack/src/main.c:1]: Reverted

File: base-hack/src/misc/file_screen.c
Comment (Sir-Chili) [reply base-hack/src/misc/file_screen.c:1]: Reverted

File: base-hack/src/misc/krusha.c
Comment (Sir-Chili) [reply base-hack/src/misc/krusha.c:1]: Well, it can help as a visual aid in first-person, as unless you have the DPad display on you have to rely on the weapon's sound to know which Kong you are, plus it is made to fit the new weapons name. I'd argue for keeping it, but it can be reverted if needed.

File: base-hack/src/pause/hints.c
Comment (Sir-Chili) [reply base-hack/src/pause/hints.c:1]: Reverted

File: randomizer/Patching/ASMPatcher.py
Comment (Sir-Chili) [reply randomizer/Patching/ASMPatcher.py:1]: Reverted

File: randomizer/Patching/CosmeticColors.py
Comment (Sir-Chili) [reply randomizer/Patching/CosmeticColors.py:2557]: Good idea, return added

File: randomizer/Patching/CosmeticColors.py
Comment (Sir-Chili) [reply randomizer/Patching/CosmeticColors.py:3587]: Reverted

File: randomizer/Patching/CosmeticColors.py
Comment (Sir-Chili) [reply randomizer/Patching/CosmeticColors.py:4210]: reverted

File: randomizer/Patching/FairyPlacer.py
Comment (Sir-Chili) [reply randomizer/Patching/FairyPlacer.py:1]: Reverted

File: randomizer/Patching/ItemRando.py
Comment (Sir-Chili) [reply randomizer/Patching/ItemRando.py:1]: Reverted

File: randomizer/Patching/UpdateHints.py
Comment (Sir-Chili) [reply randomizer/Patching/UpdateHints.py:1]: Reverted

File: base-hack/src/misc/krusha.c
Comment (AlmostSeagull) [reply base-hack/src/misc/krusha.c:1]: Fair point, keep it like this, then

File: base-hack/Build/build.py
Comment (theballaam96) [inline base-hack/Build/build.py:1105]: Remove this duplicate

**Comment by theballaam96** [comment]
To resolve the `shrink-dk64.bps` stopgap, run `./base-hack/build/refresh_static.py`, and set the branch as `dev`. That'll pull all the most recent data from dev. You then need to commit that, and then that should allow you to resolve the conflicts via the GH UI. Once you've sorted that out, rebuild the BPS.

It's annoying, but it's the only way to get around the fact that GH can't resolve conflicts with the bps file

**Comment by Sir-Chili** [comment]
There, it should be fixed to use the new and proper way to add images to ROM. It will also be disabled in case of any colorblind modes.

**Comment by theballaam96** [comment]
Still showing some merge conflict issues with the bps

**Comment by AlmostSeagull** [comment]
You'll want to merge the latest version of the dev branch into your branch to resolve conflicts (use the dev branch files for the bps and pointer addresses, and build after merging)

**Comment by AlmostSeagull** [comment]
Yeah you can merge dev into your branch and then elect to overwrite pointer_addresses with the dev version. And then build. It'll treat it as being the dev version and then edited, instead of the other way around. (same for the bps)

**Comment by Sir-Chili** [comment]
Confirmed I merged dev into my branch first, then built. I don't know if this one will work.

**Comment by AlmostSeagull** [comment]
Doesn't *look* like it. Are you sure you pulled the latest changes into your dev branch before merging it into your branch? (for me, I'd be syncing the branch on github, and then use github desktop to Pull the commits, but the exact execution of these steps vary depending on your setup)

I promise, it's easy once you get the hang of it. 

In absolute basic terms to avoid any misunderstanding and to remove the variables of what your setup is:
- You want your dev branch to have all the commits that our dev branch has.
- You then want your branch to have all the commits your dev branch now has by merging your dev branch with the branch you're working on. 
- You'll resolve any conflicts, given that you want to use the .bps and pointer_addresses files from the dev branch (actually do this for the bumpversion.cfg as well if it complains
- You now will want to build (you've probably already built during most of your attempts, but yes: again)
- And then push your changes, so the PR knows (you clearly already know this step)

**Comment by Sir-Chili** [comment]
Oh, seems like I wasn't pulling properly from the dev branch. Thank you very much for your patience and the steps, I'm not the best at handling git in general, so wrangling a couple of difficult file really is a learning experience. I should have followed all these steps properly this time.

**Comment by AlmostSeagull** [comment]
No worries, while you wait for approval for these 5 checks to start: one of them is the linting check. It checks whether or not running the lint task would change files. So you'll want to make sure you run the lint task first. Can save you some hassle (you recently merged branches so can't hurt to run it just to be sure. That's the only thing you can do to potentially speed things up here. 

I do not see any conflicts being warned about, so yeah. double check that you linted (if you use VSCode it's Terminal -> Run Task -> Lint Repo, otherwise the task is described in .vscode/tasks.json. but may be tough to figure out)

The other 4 checks tend to all succeed except for the Testing one, but that one is allowed to fail
Good luck on your wait, we'll keep you posted

**Comment by Sir-Chili** [comment]
I have run the Lint Repo task on the code in the previous commits, and it's mostly completing without errors. There's some things that it can't seem to validate since it doesn't like having python and the project on different drives, so I can't say for sure if it'll completely succeed. I'll have to wait for the check to proceed automatically it seems.

File: randomizer/Patching/CosmeticColors.py
Comment (AlmostSeagull) [inline randomizer/Patching/CosmeticColors.py:2339]: line 2339, remove all the trailing tabs/white space

File: wiki/article_markdown/custom_locations/RandomSettingsDifficult.MD
Comment (AlmostSeagull) [inline wiki/article_markdown/custom_locations/RandomSettingsDifficult.MD:98]: undo this

File: wiki/article_markdown/custom_locations/RandomSettingsDifficultWithQolShuffle.MD
Comment (AlmostSeagull) [inline wiki/article_markdown/custom_locations/RandomSettingsDifficultWithQolShuffle.MD:92]: undo this

File: wiki/article_markdown/custom_locations/RandomSettingsEasy.MD
Comment (AlmostSeagull) [inline wiki/article_markdown/custom_locations/RandomSettingsEasy.MD:109]: undo this

File: wiki/article_markdown/custom_locations/RandomSettingsStandard.MD
Comment (AlmostSeagull) [inline wiki/article_markdown/custom_locations/RandomSettingsStandard.MD:98]: undo this

**Review by AlmostSeagull — CHANGES_REQUESTED** [review]

**Comment by AlmostSeagull** [comment]
I ran the linter on a copy of your branch. Added instructions to manually apply changes the lint task would've made. These changes are required.

**Comment by theballaam96** [comment]
Approved the tests workflows. Only one of concern is the lint task, which Seagull has talked about how to amend. We'd advise using the VSCode `Lint Repo` task. Additionally, I found a duplicate of loading in the bean spin images iin `./base-hack/build.py`.

File: wiki/article_markdown/custom_locations/RandomSettingsEasy.MD
Comment (Sir-Chili) [reply wiki/article_markdown/custom_locations/RandomSettingsEasy.MD:109]: Reverting

File: wiki/article_markdown/custom_locations/RandomSettingsStandard.MD
Comment (Sir-Chili) [reply wiki/article_markdown/custom_locations/RandomSettingsStandard.MD:98]: Reverting

**Review by AlmostSeagull — DISMISSED** [review]

File: base-hack/Build/build.py
Comment (Sir-Chili) [reply base-hack/Build/build.py]: Ah, that's where the Bean Spin code that I lost went to. There was some issue where it somehow got replaced by the 25 th aniv. image loading code (as a duplicate), so I had to add it again or the pause menu would crash.

**Comment by Sir-Chili** [comment]
As I mentionned, the Lint Repo task doesn't want to play nice with the fact python and the project are on separate disks. I could move the project to the same disk to run it, however.

**Comment by AlmostSeagull** [comment]
I have python and the project on a different disk as well, it's not that. Might be that one or the other is in a folder with a space in the name or something silly like that. Either way, I told you what to change for the linting task and last time it ran, it passed the test. I'll approve the workflow again to check if the linter is happy with your changes and then see what Ballaam thinks about this

**Comment by Sir-Chili** [comment]
Ah, a stray space could also explain it. Lemme try that while the rest is processing, for fun.

**Comment by AlmostSeagull** [comment]
> Ah, a stray space could also explain it. Lemme try that while the rest is processing, for fun.

I mean, the linting test is not a problem anymore. that's solved, so better to not change anything unless Ballaam requests any changes.

**Comment by Sir-Chili** [comment]
Yeah, I'm not pushing any changes after running the "Lint Repo" command on my end, I just want to confirm it works (Which it did, it most likely was the space.)

**Review by theballaam96 — APPROVED** [review]

**Comment by theballaam96** [comment]
Looks good to me, outstanding work! Merging

**Pull request merged**
